packages/create-cloudflare/src/templateMap.ts

```diff
@@ -1,53 +1,368 @@
+	// If we can infer from the directory that it uses typescript, use that
+	if (usesTypescript(ctx)) {
+		return true;
+	}
+
+	// Otherwise, prompt the user for their TS preference
+	return processArgument<boolean>(ctx.args, "ts", {
+		type: "confirm",
+		question: "Do you want to use TypeScript?",
+		label: "typescript",
+		defaultValue: C3_DEFAULTS.ts,
+	});
+};
+
+export const processRemoteTemplate = async (args: Partial<C3Args>) => {
+	const templateUrl = await processArgument<string>(args, "template", {
+		type: "text",
+		question:
+			"What's the url of git repo containing the template you'd like to use?",
+		label: "repository",
+		validate: validateTemplateUrl,
+		defaultValue: C3_DEFAULTS.template,
+	});
+
+	const path = await downloadRemoteTemplate(templateUrl);
+	const config = inferTemplateConfig(path);
+	validateTemplate(path, config);
+
+	return {
+		path,
+		...config,
+	};
+};
+
+const validateTemplate = (path: string, config: TemplateConfig) => {
+	if (config.copyFiles?.path) {
+		validateTemplateSrcDirectory(
+			resolve(path, config.copyFiles.path as string),
+			config
+		);
+	} else {
+		for (const variant of Object.values(config.copyFiles as StaticFileMap)) {
+			validateTemplateSrcDirectory(resolve(path, variant.path), config);
+		}
+	}
+};
+
+const validateTemplateSrcDirectory = (path: string, config: TemplateConfig) => {
+	if (config.platform === "workers") {
+		const wranglerTomlPath = resolve(path, "wrangler.toml");
+		if (!existsSync(wranglerTomlPath)) {
+			crash(`create-cloudflare templates must contain a "wrangler.toml" file.`);
+		}
+	}
+
+	const pkgJsonPath = resolve(path, "package.json");
+	if (!existsSync(pkgJsonPath)) {
+		crash(`create-cloudflare templates must contain a "package.json" file.`);
+	}
+};
+
+/**
+ * Remote templates don't require a config file but may in the future. Until then, this
+ * function adapts a remote template to work with c3 by inferring a simple config from
+ * its file structure.
+ */
+const inferTemplateConfig = (path: string): TemplateConfig => {
+	return {
+		configVersion: 1,
+		id: "remote-template",
+		displayName: "A remote C3 template",
+		platform: "workers",
+		copyFiles: inferCopyFilesDefinition(path),
+	};
+};
+
+const inferCopyFilesDefinition = (path: string) => {
+	const copyFiles: StaticFileMap | VariantInfo = {};
+	if (existsSync(join(path, "js"))) {
+		copyFiles["js"] = { path: "./js" };
+	}
+	if (existsSync(join(path, "ts"))) {
+		copyFiles["ts"] = { path: "./ts" };
+	}
+	if (Object.keys(copyFiles).length !== 0) {
+		return copyFiles;
+	}
+
+	return {
+		path: ".",
+	};
+};
+
+/**
+ * Downloads an external template from a git repo using `degit`.
+ *
+ * @param src The url of the git repository to download the template from.
+ *            For convenience, `owner/repo` is also accepted.
+ * @returns A path to a temporary directory containing the downloaded template
+ */
+const downloadRemoteTemplate = async (src: string) => {
```

Comment:
If I understand correctly, we only allow templates to reside in a git repo, so there's a 1-to-1 mapping between repo and template

am I correct?

I think it would be far far more convenient to also allow a path in the repo so that a single repo can contain multiple templates
(e.g. I could have my `c3-worker-templates` repo containing a bunch of templates compared to being forced to have lots of small repos)

Could we allow such use case?
If you don't feel comfortable adding that in this PR, could you have a quick look into it and make sure that the implementation in this PR can be expanded to the use case I mentioned without the introduction of an awkward a API or a breaking change?

Reply:
I agree that this would be a good thing to consider 👍

Reply:
> am I correct?

Actually, `degit` gives us this magic out of the box 😄. For example a `src` of `cloudflare/workers-sdk/templates/workers-d1` would work with this implementation.

Reply:
Awesome, I had a feeling that was the case 😄 (from here: https://github.com/cloudflare/workers-sdk/pull/4754#discussion_r1455394448)

Thanks for confirming it 👍 

But I do feel like we should make that more clear in the yargs description if possible 🙂: https://github.com/cloudflare/workers-sdk/pull/4754/files#r1456569855